sdks/java/core/src/main/java/org/apache/beam/sdk/util/PaneInfoTracker.java

```diff
@@ -32,123 +32,206 @@
-    } else {
-      // This is the unique ON_TIME firing for the window.
-      timing = Timing.ON_TIME;
+  /**
+   * Track pane info for non-Global window.
+   */
+  private static class LocalPaneInfoTracker implements PaneInfoTrackerIface {
+
+    private TimerInternals timerInternals;
+
+    public LocalPaneInfoTracker(TimerInternals timerInternals) {
+      this.timerInternals = timerInternals;
+    }
+
+    @Override
+    public void clear(StateAccessor<?> state) {
+      state.access(PANE_INFO_TAG).clear();
     }
 
-    WindowTracing.debug(
-        "describePane: {} pane (prev was {}) for key:{}; windowMaxTimestamp:{}; "
-        + "inputWatermark:{}; outputWatermark:{}; isLateForOutput:{}",
-        timing, previousTiming, key, windowMaxTimestamp, inputWM, outputWM, isLateForOutput);
-
-    if (previousPane != null) {
-      // Timing transitions should follow EARLY* ON_TIME? LATE*
-      switch (previousTiming) {
-        case EARLY:
-          Preconditions.checkState(
-              timing == Timing.EARLY || timing == Timing.ON_TIME || timing == Timing.LATE,
-              "EARLY cannot transition to %s", timing);
-          break;
-        case ON_TIME:
-          Preconditions.checkState(
-              timing == Timing.LATE, "ON_TIME cannot transition to %s", timing);
-          break;
-        case LATE:
-          Preconditions.checkState(timing == Timing.LATE, "LATE cannot transtion to %s", timing);
-          break;
-        case UNKNOWN:
-          break;
+    /**
+     * Return a ({@link ReadableState} for) the pane info appropriate for {@code context}. The pane
+     * info includes the timing for the pane, who's calculation is quite subtle.
+     *
+     * @param isFinal should be {@code true} only if the triggering machinery can guarantee
+     *                no further firings for the
+     */
+    @Override
+    public ReadableState<PaneInfo> getNextPaneInfo(
+        ReduceFn<?, ?, ?, ?>.Context context, final boolean isFinal) {
+      final Object key = context.key();
+      final ReadableState<PaneInfo> previousPaneFuture =
+          context.state().access(PaneInfoTracker.PANE_INFO_TAG);
+      final Instant windowMaxTimestamp = context.window().maxTimestamp();
+
+      return new ReadableState<PaneInfo>() {
+        @Override
+        public ReadableState<PaneInfo> readLater() {
+          previousPaneFuture.readLater();
+          return this;
+        }
+
+        @Override
+        public PaneInfo read() {
+          return describePane(key, windowMaxTimestamp, previousPaneFuture.read(), isFinal);
+        }
+      };
+    }
+
+    @Override
+    public void storeCurrentPaneInfo(ReduceFn<?, ?, ?, ?>.Context context, PaneInfo currentPane) {
+      context.state().access(PANE_INFO_TAG).write(currentPane);
+    }
+
+    private <W> PaneInfo describePane(
+        Object key, Instant windowMaxTimestamp, @Nullable PaneInfo previousPane, boolean isFinal) {
+      boolean isFirst = previousPane == null;
```

Comment:
If you did this top level (package private), git might infer the move `PaneInfoTracker` --> `LocalPaneInfoTracker` and then view `GlobalPaneInfoTracker` as a new file. But it might insist on diffing the new and old `PaneInfoTracker`.